lib/util/extractSourceMap/helper.js

```diff
@@ -0,0 +1,357 @@
+ * @param {string} label encoding label
+ * @returns {string | null} standard encoding name or null if not found
+ */
+function labelToName(label) {
+	const labelLowercase = String(label).trim().toLowerCase();
+
+	return /** @type {EXPECTED_ANY} */ (labelsToNames)[labelLowercase] || null;
+}
+
+/**
+ * Flatten a source map that contains sections
+ * @param {RawSourceMap} map source map object
+ * @returns {Promise<RawSourceMap>} flattened source map
+ */
+async function flattenSourceMap(map) {
+	const consumer = await new sourceMap.SourceMapConsumer(map);
+	const generatedMap = map.file
+		? new sourceMap.SourceMapGenerator({
+				file: map.file
+			})
+		: new sourceMap.SourceMapGenerator();
+
+	for (const sourceFile of consumer.sources) {
+		const sourceContent = consumer.sourceContentFor(sourceFile, true);
+		generatedMap.setSourceContent(sourceFile, sourceContent);
+	}
+
+	consumer.eachMapping((mapping) => {
+		const { source } = consumer.originalPositionFor({
+			line: mapping.generatedLine,
+			column: mapping.generatedColumn
+		});
+
+		const mappings = {
+			source,
+			original: {
+				line: mapping.originalLine,
+				column: mapping.originalColumn
+			},
+			generated: {
+				line: mapping.generatedLine,
+				column: mapping.generatedColumn
+			}
+		};
+
+		if (source) {
+			generatedMap.addMapping(/** @type {Mapping} */ (mappings));
+		}
+	});
+
+	return generatedMap.toJSON();
+}
+
+/**
+ * Extract source mapping URL from code comments
+ * @param {string} code source code content
+ * @returns {SourceMappingURL} source mapping information or null
+ */
+function getSourceMappingURL(code) {
+	const lines = code.split(/^/m);
+	let match;
+
+	for (let i = lines.length - 1; i >= 0; i--) {
+		match = lines[i].match(sourceMappingURLRegex);
+		if (match) {
+			break;
+		}
+	}
+
+	const sourceMappingURL = match ? match[1] || match[2] || "" : "";
+
+	return {
+		sourceMappingURL: sourceMappingURL
+			? decodeURI(sourceMappingURL)
+			: sourceMappingURL,
+		replacementString: match ? match[0] : ""
+	};
+}
+
+/**
+ * Check if value is a URL
+ * @param {string} value string to check
+ * @returns {boolean} true if value is a URL
+ */
+function isURL(value) {
+	return /^[a-z][a-z0-9+.-]*:/i.test(value) && !path.win32.isAbsolute(value);
+}
+
+/**
+ * Get absolute path for source file
+ * @param {string} context context directory
+ * @param {string} request file request
+ * @param {string} sourceRoot source root directory
+ * @returns {string} absolute path
+ */
+function getAbsolutePath(context, request, sourceRoot) {
+	if (isURL(sourceRoot)) {
+		return new URL(request, sourceRoot).toString();
+	}
+
+	if (sourceRoot) {
+		if (path.isAbsolute(sourceRoot)) {
+			return path.join(sourceRoot, request);
+		}
+
+		return path.join(context, sourceRoot, request);
+	}
+
+	return path.join(context, request);
+}
+
+/**
+ * Fetch source content from data URL
+ * @param {LoaderContext<EXPECTED_ANY>} loaderContext loader context object
```

Comment:
I think we should not have it here, because it is not a loader. Maybe we should return this from our `extractSourceMap`, don't look deeply